rust/lance/src/datafusion/logical_expr.rs

```diff
@@ -56,60 +106,44 @@ pub fn resolve_expr(expr: &Expr, schema: &Schema) -> Result<Expr> {
                 }));
             }
             match (left.as_ref(), right.as_ref()) {
-                (Expr::Column(l), Expr::Literal(_)) => {
-                    let Some(field) = schema.field(&l.flat_name()) else {
-                        return Err(Error::IO {
-                            message: format!(
-                                "Column {} does not exist in the dataset.",
-                                l.flat_name()
-                            ),
-                            location: location!(),
-                        });
-                    };
-                    Ok(Expr::BinaryExpr(BinaryExpr {
-                        left: left.clone(),
-                        op: *op,
-                        right: Box::new(resolve_value(right.as_ref(), &field.data_type())?),
-                    }))
+                (Expr::Column(_) | Expr::GetIndexedField(_), Expr::Literal(_)) => {
+                    if let Some(resolved_type) = resolve_column_type(left.as_ref(), schema) {
+                        Ok(Expr::BinaryExpr(BinaryExpr {
+                            left: left.clone(),
+                            op: *op,
+                            right: Box::new(resolve_value(right.as_ref(), &resolved_type)?),
+                        }))
+                    } else {
+                        Ok(expr.clone())
+                    }
                 }
-                (Expr::Literal(_), Expr::Column(l)) => {
-                    let Some(field) = schema.field(&l.flat_name()) else {
-                        return Err(Error::IO {
-                            message: format!(
-                                "Column {} does not exist in the dataset.",
-                                l.flat_name()
-                            ),
-                            location: location!(),
-                        });
-                    };
-                    Ok(Expr::BinaryExpr(BinaryExpr {
-                        left: Box::new(resolve_value(right.as_ref(), &field.data_type())?),
-                        op: *op,
-                        right: right.clone(),
-                    }))
+                (Expr::Literal(_), Expr::Column(_) | Expr::GetIndexedField(_)) => {
+                    if let Some(resolved_type) = resolve_column_type(right.as_ref(), schema) {
+                        Ok(Expr::BinaryExpr(BinaryExpr {
```

Comment:
This looks like a slight behavior change when `Column` references a field that doesn't exist.  Before we were returning an error.  Now, `resolve_column_type` will return `None` and so we return the original expression.

Is it just that we should never get into this situation to begin with?  Or is is that we're going to throw an error later anyways so there is no real need to detect it right now?

Reply:
Good question. Basically this will be caught later by other passes. We could add error handling here, but it feels unnecessary.